src/libraries/System.Net.Security/src/System/Net/Security/TlsFrameHelper.cs

```diff
@@ -179,27 +179,51 @@ public static bool TryGetFrameHeader(ReadOnlySpan<byte> frame, ref TlsFrameHeade
             {
                 header.Type = (TlsContentType)frame[0];
 
-                if (frame.Length >= 3)
+                if (frame.Length > 4)
                 {
                     // SSLv3, TLS or later
                     if (frame[1] == 3)
                     {
-                        if (frame.Length > 4)
-                        {
-                            header.Length = ((frame[3] << 8) | frame[4]);
-                        }
-
+                        header.Length = ((frame[3] << 8) | frame[4]);
                         header.Version = TlsMinorVersionToProtocol(frame[2]);
+                        return true;
                     }
-                    else
+                    else if (frame[2] == (byte)TlsHandshakeType.ClientHello &&
+                       frame[3] == 3) // SSL3 or above
                     {
-                        header.Length = -1;
-                        header.Version = SslProtocols.None;
+                        int length;
+                        if ((frame[0] & 0x80) != 0)
+                        {
+                            // Two bytes
+                            length = (((frame[0] & 0x7f) << 8) | frame[1]) + 2;
+                        }
+                        else
+                        {
+                            // Three bytes
+                            length = (((frame[0] & 0x3f) << 8) | frame[1]) + 3;
+                        }
+
+                        // max frame for SSLv2 is 32767.
+                        // However, we expect something reasonable for initial HELLO
+                        // We don't have enough logic to verify full validity,
+                        // the limits bellow are queses.
+                        if (length > 20 && length < 1000)
```

Comment:
The intention here was to check that the length is "reasonable" for hello. I don't know if we need to. In YARP this can be used to eliminate garbage without even allocating SslStream. We don't do that here but it would be nice IMHO to have both instances same (or close as possible)